### PR TITLE
[soundcloud] Enriched search results

### PR DESCRIPTION
**Pull request by SuperSonicHub1**
### Description of your *pull request* and other information

</details>

Moved _extract_info_dict, _IMAGE_REPL_RE, _ARTWORK_MAP from SoundCloudIE to SoundcloudBaseIE so that they could be used in any subextractor. Modified _extract_info_dict to take an 'extract_flat' argument. When false, it will still extract info from a SoundCloud track, but it will not make any network requests. Used _extract_info_dict(extract_flat=False) to then fatten up search results of SoundcloudSearchIE. I also fixed SoundcloudSearchIE's test case by adding the 'id' field.

Preview:
```json
{
  "id": "hello",
  "title": "hello",
  "_type": "playlist",
  "entries": [
    {
      "id": "230155983",
      "uploader": "Grace Davies",
      "uploader_id": "44817748",
      "uploader_url": "https://soundcloud.com/gracedaviesofficial",
      "timestamp": 1445871069,
      "title": "Hello - Adele",
      "description": "Vocals/Piano/Recording/Production by me, in my bedroom!\n\nFollow me on:\nWebsite: www.gracedaviesofficial.com\nFacebook: gracedaviesofficial\nYouTube: gracedaviesoffical\nInstagram: gracedaviesofficial\nSnapchat: gracedofficial\nTwitter: gracedavies",
      "thumbnails": [
        {
          "id": "mini",
          "url": "https://i1.sndcdn.com/artworks-000134001026-tugyco-mini.jpg",
          "width": 16,
          "height": 16
        },
        {
          "id": "tiny",
          "url": "https://i1.sndcdn.com/artworks-000134001026-tugyco-tiny.jpg",
          "width": 20,
          "height": 20
        },
        {
          "id": "small",
          "url": "https://i1.sndcdn.com/artworks-000134001026-tugyco-small.jpg",
          "width": 32,
          "height": 32
        },
        {
          "id": "badge",
          "url": "https://i1.sndcdn.com/artworks-000134001026-tugyco-badge.jpg",
          "width": 47,
          "height": 47
        },
        {
          "id": "t67x67",
          "url": "https://i1.sndcdn.com/artworks-000134001026-tugyco-t67x67.jpg",
          "width": 67,
          "height": 67
        },
        {
          "id": "large",
          "url": "https://i1.sndcdn.com/artworks-000134001026-tugyco-large.jpg",
          "width": 100,
          "height": 100
        },
        {
          "id": "t300x300",
          "url": "https://i1.sndcdn.com/artworks-000134001026-tugyco-t300x300.jpg",
          "width": 300,
          "height": 300
        },
        {
          "id": "crop",
          "url": "https://i1.sndcdn.com/artworks-000134001026-tugyco-crop.jpg",
          "width": 400,
          "height": 400
        },
        {
          "id": "t500x500",
          "url": "https://i1.sndcdn.com/artworks-000134001026-tugyco-t500x500.jpg",
          "width": 500,
          "height": 500
        },
        {
          "id": "original",
          "url": "https://i1.sndcdn.com/artworks-000134001026-tugyco-original.jpg",
          "preference": 10
        }
      ],
      "duration": 300.88,
      "webpage_url": "https://soundcloud.com/gracedaviesofficial/hello-adele",
      "license": "all-rights-reserved",
      "view_count": 67911929,
      "like_count": 1074562,
      "comment_count": 8909,
      "repost_count": 53298,
      "genre": "hello",
      "formats": [],
      "ie_key": "Soundcloud",
      "_type": "url",
      "url": "https://api.soundcloud.com/tracks/230155983",
      "__x_forwarded_for_ip": null
    }
  ],
  "webpage_url": "scsearch1:hello",
  "original_url": "scsearch1:hello",
  "webpage_url_basename": "hello",
  "webpage_url_domain": null,
  "extractor": "soundcloud:search",
  "extractor_key": "SoundcloudSearch",
  "requested_entries": [
    1
  ],
  "playlist_count": 1,
  "__files_to_move": {},
  "epoch": 1663544663,
  "_version": {
    "version": "2022.09.01",
    "current_git_head": "b2d5beeaa",
    "release_git_head": "5d7c7d656",
    "repository": "yt-dlp/yt-dlp"
  }
}
```

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [X] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [X] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [X] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [X] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [X] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

**Comment by pukkandan** [comment]
1. When flat, `formats` must return `None`, not `[]`
2. You got the param name reversed? Currently the extraction is flat when `extract_flat=False`

**Comment by SuperSonicHub1** [comment]
1. Fixed it!
2. Whoops, mixed it up in my head.

By the way, sorry for taking so long to sit down for a one-minute patch. Life always has to get in the way...

File: yt_dlp/extractor/soundcloud.py [hunk 1]
```diff
@@ -179,6 +194,158 @@ def sign(self, user, pw, clid):
...
+        transcodings = try_get(
+            info, lambda x: x['media']['transcodings'], list) or []
+        for t in transcodings:
+            if not isinstance(t, dict):
+                continue
+            format_url = url_or_none(t.get('url'))
+            if not format_url:
+                continue
+            stream = self._download_json(
+                format_url, track_id, query=query, fatal=False, headers=self._HEADERS) if not extract_flat else None
```
Comment (pukkandan) [inline yt_dlp/extractor/soundcloud.py:271]:
```suggestion
            stream = None if extract_flat else self._download_json(
                format_url, track_id, query=query, fatal=False, headers=self._HEADERS)
```

**Review by pukkandan — APPROVED** [review]

**Comment by SuperSonicHub1** [comment]
Glad I could be of use! More pulls should hopefully be on the way!

**Comment by SuperSonicHub1** [comment]
@pukkandan I think we need a follow up on this; yt-dlp does not like formats being `None`:
```
$ yt-dlp -J --flat-playlist -v scsearch:love
[debug] Command-line config: ['-J', '--flat-playlist', '-v', 'scsearch:love']
[debug] User config "/home/kwilliams/.config/yt-dlp/config": ['--netrc', '--sub-lang', 'en,en-US,eng', '--format', '(bestvideo+bestaudio/best)[format_id*=en]/(bestvideo+bestaudio/best)']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.10.04 [4e0511f] (pip) API
[debug] Python 3.10.0 (CPython 64bit) - Linux-5.15.0-48-generic-x86_64-with-glibc2.35 (glibc 2.35)
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] exe versions: ffmpeg 4.4.2 (setts), ffprobe 4.4.2, rtmpdump 2.4
[debug] Optional libraries: Cryptodome-3.15.0, brotli-1.0.9, certifi-2022.09.24, mutagen-1.45.1, sqlite3-2.6.0, websockets-10.3
[debug] Proxy map: {}
[debug] Loaded 1690 extractors
WARNING: [soundcloud:search] parsing .netrc: No authenticators for soundcloud (None, line None)
[debug] [soundcloud:search] Extracting URL: scsearch:love
[download] Downloading playlist: love
[soundcloud:search] love: Downloading page 1
[soundcloud:search] Playlist love: Downloading 1 videos of 1
[debug] The information of all playlist entries will be held in memory
[download] Downloading video 1 of 1
ERROR: 'NoneType' object is not iterable
Traceback (most recent call last):
  File "/home/kwilliams/.cache/pypoetry/virtualenvs/playah-6OEbIXqo-py3.10/lib/python3.10/site-packages/yt_dlp/YoutubeDL.py", line 1477, in wrapper
    return func(self, *args, **kwargs)
  File "/home/kwilliams/.cache/pypoetry/virtualenvs/playah-6OEbIXqo-py3.10/lib/python3.10/site-packages/yt_dlp/YoutubeDL.py", line 1876, in __process_iterable_entry
    return self.process_ie_result(
  File "/home/kwilliams/.cache/pypoetry/virtualenvs/playah-6OEbIXqo-py3.10/lib/python3.10/site-packages/yt_dlp/YoutubeDL.py", line 1623, in process_ie_result
    info_copy, _ = self.pre_process(info_copy)
  File "/home/kwilliams/.cache/pypoetry/virtualenvs/playah-6OEbIXqo-py3.10/lib/python3.10/site-packages/yt_dlp/YoutubeDL.py", line 3438, in pre_process
    info = self.run_all_pps(key, info)
  File "/home/kwilliams/.cache/pypoetry/virtualenvs/playah-6OEbIXqo-py3.10/lib/python3.10/site-packages/yt_dlp/YoutubeDL.py", line 3429, in run_all_pps
    self._forceprint(key, info)
  File "/home/kwilliams/.cache/pypoetry/virtualenvs/playah-6OEbIXqo-py3.10/lib/python3.10/site-packages/yt_dlp/YoutubeDL.py", line 2818, in _forceprint
    info_copy['formats_table'] = self.render_formats_table(info_dict)
  File "/home/kwilliams/.cache/pypoetry/virtualenvs/playah-6OEbIXqo-py3.10/lib/python3.10/site-packages/yt_dlp/YoutubeDL.py", line 3605, in render_formats_table
    table = [
TypeError: 'NoneType' object is not iterable

[download] Finished downloading playlist: love
{"id": "love", "title": "love", "_type": "playlist", "entries": [null], "webpage_url": "scsearch:love", "original_url": "scsearch:love", "webpage_url_basename": "love", "webpage_url_domain": null, "extractor": "soundcloud:search", "extractor_key": "SoundcloudSearch", "requested_entries": [1], "playlist_count": 1, "__files_to_move": {}, "epoch": 1664923353, "_version": {"version": "2022.10.04", "current_git_head": null, "release_git_head": "4e0511f", "repository": "yt-dlp/yt-dlp"}}
```

What I don't understand is why we are even in `render_formats_table` even though the argument isn't set.

**Pull request merged**